ISPN-8865 Move AdvancedCacheLoader over to using Publisher

wburns opened the pull request:
https://issues.jboss.org/browse/ISPN-8865

Added AdvancedCacheLoader.publishEntries method and deprecated/removed all references to calling old process method that I could.

This JIRA is mostly to let CI run on it and crunch out some stuff to see if there are test failures so far.

wburns commented:
All current loaders have been converted except the SoftIndexFileStore, which I will tackle next week.

rvansa commented on core/src/main/java/org/infinispan/filter/KeyFilter.java:
```diff
@@ -10,8 +10,10 @@
+ * @deprecated This will be replaced by {@link java.util.function.Predicate} in the future
  */
 @ThreadSafe
+@Deprecated
```
Why don't you let this already extend Predicate? That way you could safely change the method parameters to Predicate because all KeyFilters passed in would be Predicates...

wburns replied:
Sure, I am not sure why I didn't do this before 👍

rvansa commented on core/src/main/java/org/infinispan/persistence/PersistenceUtil.java:
```diff
@@ -37,26 +38,30 @@ public static KeyFilter notNull(KeyFilter filter) {
-      final AtomicInteger result = new AtomicInteger(0);
-      acl.process(filter, (marshalledEntry, taskContext) -> result.incrementAndGet(), new WithinThreadExecutor(), false, false);
-      return result.get();
+   public static <K, V> int count(AdvancedCacheLoader<K, V> acl, Predicate<? super K> filter) {
```
This class is public, you need to make the change backwards-compatible.

wburns replied:
above will fix that

rvansa commented on core/src/main/java/org/infinispan/persistence/PersistenceUtil.java:
```diff
@@ -37,26 +38,30 @@ public static KeyFilter notNull(KeyFilter filter) {
-      final Set<K> set = new HashSet<K>();
-      acl.process(filter, (marshalledEntry, taskContext) -> set.add(marshalledEntry.getKey()), new WithinThreadExecutor(), false, false);
-      return set;
+      return Flowable.fromPublisher(acl.publishEntries(filter, false, false))
```
would a separate publishKeys method make sense?

wburns replied:
I was actually thinking about that over the weekend. I was debating about having a new method or possibly adding a "builder" API (but this might be overkill).

Something like

```
AdvancedCacheLoader {
   ...
   StorePublisher publish();
   ...
}

interface StorePublisher {
   Publisher<K> keys();

  Publisher<MarshalledEntry<K, V>> entries(boolean fetchValues, boolean fetchKeys);
}
```

What do you think?

rvansa replied:
I wouldn't scope it under such interface, just
```
Publisher<K> publishKeys(Predicate<? super K> filter);
Publisher<MarshalledEntry<K, V>> publishEntries(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata);
```
I have considered dropping the `fetchValue` completely but there might be use case for key + metadata (e.g. expiration).

wburns replied:
Yeah I am fine with the methods as well - and obviously I can make it a nice default method :)

I am thinking that `publishKeys` will have to be documented saying it is up to the cache store implementation whether or not expired keys are returned.

In regards to `fetchValue` I can look over the usages again and see if we can possibly drop it. Although to be honest with you most implementations if you supply either `fetchValue` or `fetchMetadata` it is just as easy to get the other. In that case I wonder if it makes any sense to have these arguments anymore.

wburns replied:
I guess scattered cache requires just the metadata :) So I guess I will keep both arguments for now.

wburns replied:
Actually looking at this more `publishKeys` has to guarantee that it returns non expired keys. Or else the usefulness of the method is quite questionable (for example keySet iterator can't use it otherwise). So in that case I will try to limit perf impact as much as possible.

rvansa commented on core/src/main/java/org/infinispan/util/Closeables.java:
```diff
@@ -0,0 +1,38 @@
+
+/**
+ * @author wburns
+ * @since 9.0
```
Bah, another Closeables class... btw. the since tag does not match.

wburns replied:
Yeah unfortunately reactive streams isn't in commons yet :(

My eclipse isn't updated - thanks 👍

rvansa replied:
Would it hurt to rather add the dependency there? We want to eventually keep up parity with remote clients anyway...

wburns replied:
Afaik it shouldn't hurt. I agree, I was trying to delay the introduction of it until we absolutely needed it in the client.

wburns commented:
I have added in SoftIndexFileStore support. This also fixes https://issues.jboss.org/browse/ISPN-8880

I have also updated all the various stores to have the new `publishKeys` method. Those that don't implement this gain no real benefit so they just fall back to default.

Also I made sure to go through and make sure that all stores don't return expired entries/keys. There were a couple bugs around this as reactive streams spec doesn't allow for null values and thus a couple map methods had to be redone.

wburns commented:
Also I am fine with reactive streams int commons, but I think we need to discuss further before moving rxjava2. In this case I can't remove the Closeables class yet.

So that still leaves me to refactor the RestStore so it does async which will allow for better performance. I am thinking of capping the number of concurrent async gets. This may be configurable or just a sensible # like number of CPUs.

So the other only change at this time I can think of is possibly adding a parallel observation of JpaStore elements when `fetchValues` or `fetchMetadata` is true since this will cause a lot of blocking for the requesting thread.

wburns commented:
Updated and cleaned up. Should be ready for 9.3 after we branch now.

wburns commented:
I am still not keen on using the persistence executor for the JPAStore or LuceneCacheLoader as this only has 4 threads by default. I could possibly use the `Schedulers.io` but this would be a different thread pool.

wburns commented:
There are test failures: https://ci.infinispan.org/job/Infinispan/job/PR-5794/6/#showFailuresLink
Also I need to add reactive streams to the parent dependencies for cache stores it seems, have to look closer.

wburns commented:
I have rebased on top of https://github.com/infinispan/infinispan/pull/5824 now as well. I still have to clean up test failures.

wburns commented:
https://ci.infinispan.org/job/Infinispan/job/PR-5794/7/#showFailuresLink

I have fixed 2, but for whatever reason I can't reproduce the various class loading issue ones. I will have to look at it closer next week.

wburns commented:
Added modules, however I think there is at least 1 or more failure I still need to fix for Monday.

wburns commented:
Closing this until https://github.com/infinispan/infinispan/pull/5824 is integrated.

danberindei commented:
Please reopen Will, I didn't realize most changes were in #5824, so I have lots of pending comments here :)

wburns commented:
Reopening, this is missing a few updates, but I can do that after opened.

wburns commented on core/src/main/java/org/infinispan/cache/impl/EncoderCache.java:
```diff
@@ -221,6 +229,29 @@ public boolean contains(Object o) {
+      Collection<K> buildKeyCollection(Collection<?> c) {
+         Collection<K> toRemove = new ArrayList<>(c.size());
+         for (Object obj : c) {
+            K entry = keyToStorage(obj);
```
So I was debating about both these points a bit.

I think I need to go back and just implement all the methods and let the underlying sets do their jobs. I should make a JIRA to do this though since it isn't directly related to the future changes here.

wburns replied:
I created https://issues.jboss.org/browse/ISPN-8946

wburns commented on core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java:
```diff
@@ -491,23 +455,21 @@ public WrappedEntrySet(EntrySetCommand command, CacheSet<CacheEntry<K, V>> entry
-         CloseableIterator<CacheEntry<K, V>> iterator = Closeables.iterator(entrySet.stream());
+         // This can be a HashSet since it is only written to from the local iterator which is only invoked
+         // from user thread
          Set<K> seenKeys = new HashSet<>(cache.getAdvancedCache().getDataContainer().sizeIncludingExpired());
```
1. I don't know if the listener will work very well. I will have to listen to all modifications to the cache as well all activations and passivations. My thought for fixing this was to have it store all if it is OBJECT storage and only a subset if it is OFF-HEAP or BINARY.

2. True about the injection, I can do that.

wburns replied:
Also a side note but after I add in segmented store and data container, this code will only be used for non dist stores. And it is possible with other cache modes to do iteration over each segment individually so that you only have so many resurrected objects.

But I will think more on doing a live view, it gets a bit messy with the listeners.

https://github.com/wburns/infinispan/blob/ISPN-8905_non_shared_store_segments/core/src/main/java/org/infinispan/stream/impl/local/PersistencKeyStreamSupplier.java#L78

wburns commented on core/src/main/java/org/infinispan/persistence/PersistenceUtil.java:
```diff
@@ -37,26 +38,73 @@ public static KeyFilter notNull(KeyFilter filter) {
+    * @param <K>
+    * @param <V>
+    * @return
+    * @deprecated Please use {@link #toKeySet(AdvancedCacheLoader, Predicate)} instead
```
agreed

wburns commented on core/src/main/java/org/infinispan/persistence/async/AdvancedAsyncCacheLoader.java:
```diff
@@ -1,120 +1,84 @@
+   public Publisher<MarshalledEntry<K, V>> publishEntries(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata) {
+      State state = this.state.get();
+      ByRef<Boolean> hadClear = new ByRef<>(Boolean.FALSE);
+      Map<Object, Modification> modificationMap = state.flattenModifications(hadClear);
```
Unfortunately we need to make a copy, otherwise the state can be updated after iterating through it and causing us to exclude an entry from the store that we shouldn't have.

wburns replied:
Yep my thoughts as well.

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -488,47 +488,54 @@ public boolean delete(Object key) {
+         synchronized (entries) {
+            for (Map.Entry<K, FileEntry> e : entries.entrySet()) {
+               K key = e.getKey();
+               if ((filter == null || filter.test(key)) && !e.getValue().isExpired(now)) {
```
Sure

wburns commented on core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java:
```diff
@@ -488,47 +488,54 @@ public boolean delete(Object key) {
+   public Flowable<MarshalledEntry<K, V>> publishEntries(Predicate<? super K> filter, boolean fetchValue, boolean fetchMetadata) {
+      if (fetchMetadata || fetchValue) {
+         return Flowable.fromIterable(() -> {
+            // This way the sorting of entries is lazily done on each invocation
```
the publisher :)

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java:
```diff
@@ -57,13 +58,46 @@
+   /**
+    * Returns a publisher that will publish all entries stored by the underlying cache store. Only the first
+    * cache store that implements {@link AdvancedCacheLoader} will be used. Predicate is applied by the underlying
+    * loader in a best attempt to improve performance.
```
If it doesn't implement `AdvancedCacheLoader` we can't use it since it doesn't implement the needed method. This is how it used to work before, the big change now is we don't invoke the publisher on _every_ store that implements `AdvancedCacheLoader` which to me was a big bug.

wburns replied:
Hrmm, I worry about duplicates with doing this, but I can add back in supporting more than 1 loader.

wburns replied:
I don't think we can put the requirement that you have to implement `CacheWriter` if you implement `AdvancedCacheLoader`. I can see a user possibly wanting a read only backed store that doesn't allow writes. This is a use case of invalidation imo. To be honest `JPAStore` also falls into this niche (it probably should have had a read only and read/write variant tbh).

Also I still really don't like the idea of loading from more than 1 cache loader. We will almost certainly have duplicates, and there is no reasonable way we can limit the memory usage to filter said duplicates out (segments will help but still have issues most likely).

danberindei replied:
Looks like I edited my comment too much and I removed the part where I was agreeing with you that we should only load from one `AdvancedCacheLoader`. I was mentioning `CacheWriter` because 2 loaders that are also writers *should* have the same entries, so there's no benefit in loading entries for both there. But TBH I never understood the use case for using multiple writers on the same cache :)

Anyway, I think restricting publishKeys/Entries to a single `AdvancedCacheLoader` is a good idea, we just need to make the user aware of it.

wburns replied:
I created https://issues.jboss.org/browse/ISPN-9144

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -246,22 +247,16 @@ public void preload() {
-         }
-         Metadata metadata = me.getMetadata() != null ? ((InternalMetadataImpl) me.getMetadata()).actual() :
+      Long insertAmount = Flowable.fromPublisher(preloadCl.publishEntries(null, true, true))
+         .take(maxEntries).doOnNext(me -> {
```
Sure

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java:
```diff
@@ -246,22 +247,16 @@ public void preload() {
+      Long insertAmount = Flowable.fromPublisher(preloadCl.publishEntries(null, true, true))
+         .take(maxEntries).doOnNext(me -> {
+            Metadata metadata = me.getMetadata() != null ? ((InternalMetadataImpl) me.getMetadata()).actual() :
                null; //the downcast will go away with ISPN-3460
```
Yeah I agree, I just kinda left it alone - but I will clean it up.

wburns commented on core/src/main/java/org/infinispan/persistence/spi/InitializationContext.java:
```diff
@@ -35,4 +37,11 @@
+
+   /**
+    * Returns the preferred executor to be used by stores if needed. Stores normally shouldn't need this unless they
+    * *must* perform some blocking code synchronously.
```
Rereading this, I didn't finish my sentence. But, saying asynchronously instead makes it less confusing, than my actual sentence would have been 👍

wburns commented on core/src/main/java/org/infinispan/persistence/spi/AdvancedCacheLoader.java:
```diff
@@ -68,8 +121,10 @@
    /**
     * Used during the parallel iteration in order to offer the {@link CacheLoaderTask} a way of canceling the entry
     * iteration. Should be thread safe.
+    * @deprecated The process method is no longer suggested and thus this class shouldn't be in use any more
```
Okay I can add since 9.3. I am hoping to remove all these guys in 10.0 :)

wburns commented on core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java:
```diff
@@ -83,15 +82,16 @@
 /**
  * Implementation of {@link CacheStream} that provides support for lazily distributing stream methods to appropriate
  * nodes
+ * @param <Original> the original type of the underlying stream - normally CacheEntry or Object
```
The type is exposed only for some type checks. To be honest the typing is not completely done here as the IntermediateOperation while having types on them are not provided (since when I add them I didn't have a good way to do it) I do now, but it requires more changes I didn't want to put in here yet.

https://github.com/infinispan/infinispan/pull/5794/files#diff-c37d0eb44c3ac6aad4102f7b81ec346dR278
https://github.com/infinispan/infinispan/pull/5794/files#diff-c37d0eb44c3ac6aad4102f7b81ec346dR129

wburns commented on core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java:
```diff
@@ -455,7 +431,8 @@ public long count() {
       } else {
-         Iterable<IntermediateOperation> ops = iteratorOperation.prepareForIteration(intermediateOperations);
+         Iterable<IntermediateOperation> ops = iteratorOperation.prepareForIteration(intermediateOperations,
+               (Function) nonNullKeyFunction());
```
Unfortunately enums don't like type parameters and there is no way to not get this to complain, so I just took the route of less code :)

wburns commented on core/src/main/java/org/infinispan/util/DoubleIterator.java:
```diff
@@ -0,0 +1,74 @@
+ * second iterator is consumed this iterator will also be completed.
+ * @param <E> element type
+ */
+public class DoubleIterator<E> implements CloseableIterator<E> {
```
Sure

wburns commented on core/src/main/java/org/infinispan/stream/impl/local/EntryStreamSupplier.java:
```diff
@@ -42,18 +42,22 @@ public EntryStreamSupplier(Cache<K, V> cache, boolean remoteIterator, ToIntFunct
    @Override
    public Stream<CacheEntry<K, V>> buildStream(Set<Integer> segmentsToFilter, Set<?> keysToFilter) {
       Stream<CacheEntry<K, V>> stream;
+      // Make sure we aren't going remote to retrieve these
```
I am thinking to change this to "Make sure we read unencoded entries (intermediate operations handle this) and don't go retrieve remote entries as rehash logic will handle this"

wburns commented on core/src/main/java/org/infinispan/util/DoubleIterator.java:
```diff
@@ -0,0 +1,74 @@
+      if (iterator2.hasNext()) {
+         iterator2Next = iterator2.next();
+      }
+      return iterator2Next != null;
```
Hrmm I agree about null. To be honest this class was based off of a previous one that required objects to be non null. However there is no reason this one can't allow null values. I can fix this.

Actually tbh with you I think the whole iteration process will have lots of trouble returning null values (reactive stream spec for example does not allow null values). Normally you can't get a null value from a ConcurrentMap, but unfortunately if you use a stream with a map operator you can get an iterator that returns null there.

I am not sure if we want to start supporting null all the way down the stack or not. It will require most likely a lot of changes.

wburns commented on core/src/test/java/org/infinispan/persistence/ParallelIterationTest.java:
```diff
@@ -136,39 +143,61 @@ private void runIterationTest(Executor executor, final boolean fetchValues,
+      flowable.subscribe(subscriber);
+
+      int batchsize = 10;
+      // Request over the max amount so we can ensure the subscriber will get notified of completion
```
the comment is a little outdated, but we request all entries except the last batch (although if the # of entries was a multiple of 9 it wouldn't do that correctly).

The reason this is done is because the way this work is the first thread to request the entries could return all the entries (depends on fast/slow it is running) and this allows me to request at least the last batch on the main thread - guaranteeing we have processing of entries on at least 2 threads.

wburns commented on core/src/test/java/org/infinispan/persistence/ParallelIterationTest.java:
```diff
@@ -136,39 +143,61 @@ private void runIterationTest(Executor executor, final boolean fetchValues,
+      int batchsize = 10;
+      // Request over the max amount so we can ensure the subscriber will get notified of completion
+      for (int i = 0; i < NUM_ENTRIES / batchsize - 1; i++) {
+         // Now request all the entries on different threads - just to see if the publisher can handle it
```
See other comment it is a bit outdated and purposely isn't requesting all entries.

wburns commented on core/src/test/java/org/infinispan/persistence/ParallelIterationTest.java:
```diff
@@ -136,39 +143,61 @@ private void runIterationTest(Executor executor, final boolean fetchValues,
+         executor.execute(() -> subscriber.request(batchsize));
+      }
+
+      subscriber.awaitCount(NUM_ENTRIES - 10, () -> {
```
yes, it is the batch size. I can update to have the variable.

wburns commented on persistence/jdbc/src/main/java/org/infinispan/persistence/jdbc/stringbased/JdbcStringBasedStore.java:
```diff
@@ -45,28 +46,24 @@
- * <p/>
- * <b>Preload</b>.In order to support preload functionality the store needs to read the string keys from the database and transform them
- * into the corresponding key objects. {@link org.infinispan.persistence.keymappers.Key2StringMapper} only supports
+ * <b>Preload & Iteration</b>.In order to support preload or iteration functionality the store needs to read the string
```
Sure

wburns commented on lucene/lucene-directory/src/main/java/org/infinispan/lucene/cacheloader/LuceneCacheLoader.java:
```diff
@@ -91,38 +92,20 @@ public boolean contains(final Object key) {
+               return Flowable.fromIterable(allInternalEntries);
+            })
+            .filter(me -> filter == null || filter.test(me.getKey()))
+            .sequential();
```
The user can still do whatever they want to the resulting publisher here. The problem is with reactive streams is you can't call `onNext` from different threads concurrently. This code here is ensuring that `onNext` observes this contract - note it will most likely be called from different threads, just not at the same time.

The user can still observe these entries on any number of threads that they want. Note a publisher is not "used" until someone subscribes to it.

wburns commented on core/src/test/java/org/infinispan/persistence/ParallelIterationTest.java:
```diff
@@ -136,39 +143,61 @@ private void runIterationTest(Executor executor, final boolean fetchValues,
+      }, TimeUnit.SECONDS.toMillis(10));
+
+      // Now request on the main thread - which should guarantee requests from different threads
+      subscriber.request(11);
```
So I think I just need to cleanup the comments here. 

First the size of 11 is here just as a further test (there should only be 10 entries, but I am making sure I only get back 10 despite requesting 11).

And actually the key part is that it is guaranteeing that objects are returned via `onNext` on different threads. The request is already guaranteed to be on different threads from above, but for performance reasons some publishers might only call `onNext` on one of those threads (ones that add entries to a queue for example (and it is faster to have one thread drain all elements from the queue before finishing).

For example our ClusterStreamManagerImpl publisher only allows for 1 remote request per node to guarantee and adds additional requests (so future requests can request all of those) until any pending request is complete.

wburns commented:
Updated and rebased, need to ensure tests still pass and add documentation though.

wburns commented:
There appear to be some related test failures https://ci.infinispan.org/job/Infinispan/job/PR-5794/18/

wburns commented:
Tests should be fixed now, waiting on latest run.

danberindei commented on persistence/jpa/src/main/java/org/infinispan/persistence/jpa/JpaStore.java:
```diff
@@ -36,25 +37,29 @@
 /**
- *
+ * NOTE: This store can return expired entries on any given operation if {@link JpaStoreConfiguration#storeMetadata()}
+ * was not set to true.
```
Does that include `publishKeys`? It doesn't really return entries, but I don't see any additional checks.

danberindei replied:
Actually I'd like you to change the javadoc of `publishKeys` in `AdvancedCacheLoader` and `PersistenceManager`, because they say "This publisher will never return a key which belongs to an expired entry"

wburns replied:
Hrmm well that is the contract, which I would like to keep on `AdvancedCacheLoader`. JpaStore is special in that it violates this contract, which is why I would rather put it here. I don't want an implementation to return expired keys. 

For this reason DummyStore doesn't implement publishKeys, because the metadata is stored with entry so we have to read that as well.

danberindei replied:
I don't think a javadoc here is enough, because there's nothing anyone can do about `JpaStore` breaking the contract:
* We won't add a special case for `JpaStore` in core, we're going to call `publishKeys` based on the `AdvancedCacheLoader` javadoc.
* Users call methods on the cache, they don't necessarily know which cache method calls which store method internally.

IMO we should throw an exception when the user tries to store mortal entries in a `JpaStore` if we can't read the metadata back. It's a bit complicated by the async writer catching all the exceptions, but at least they'll see some errors in the log. And ideally I'd like to make expiration support optional, both at the store level and at the cache level, so we can throw that exception right from `CacheImpl`.

wburns replied:
Hrmm I think it might be easier to log another JIRA for this.

It would cover

1. I would like to throw an exception at start up if JPAStore is configured not being able to store metadata. Unfortunately we can't do this, since they could be using the store as a read only.
2. We probably can only throw exceptions at runtime if an entry is written to the JPAStore with expiration.

Either way neither of these are great options, thus why I would recommend a separate JIRA to talk about this.

wburns replied:
I created https://issues.jboss.org/browse/ISPN-9143 to continue this.

danberindei commented on persistence/rest/src/main/java/org/infinispan/persistence/rest/RestStore.java:
```diff
@@ -58,18 +56,20 @@
  *
  * @author Tristan Tarrant
  * @since 6.0
+ * @deprecated This cache store will be changed to only implement {@link org.infinispan.persistence.spi.CacheLoader}
```
Aren't we using it for rolling upgrades?

tristantarrant replied:
Nope, Hot Rod only

wburns commented on core/src/main/java/org/infinispan/filter/KeyFilter.java:
```diff
@@ -10,9 +12,11 @@
  *
  * @author Manik Surtani
  * @since 6.0
+ * @deprecated This will be replaced by {@link java.util.function.Predicate} in the future
```
Sure, forgot to put one here.

wburns commented on core/src/main/java/org/infinispan/interceptors/impl/CacheLoaderInterceptor.java:
```diff
@@ -534,15 +498,16 @@ public int size() {
-         }
-         return empty;
-      }
+      // TODO: this can't be added in until stores size method is guaranteed to not count expired entries
```
Sure sounds good. I had thought about doing that with iterator, but didn't think of `publishKeys`. I think it has just been too long since i looked at this.

wburns commented on core/src/main/java/org/infinispan/persistence/manager/PersistenceManager.java:
```diff
@@ -56,13 +57,46 @@
+    * loader in a best attempt to improve performance.
+    * <p>
+    * Caller can tell the store to also fetch the value or metadata. In some cases this can improve performance. If
+    * metadata is not fetched it is not guaranteed if the publisher will return expired entries or not.
```
Okay

wburns commented on core/src/main/java/org/infinispan/persistence/spi/AdvancedCacheLoader.java:
```diff
@@ -47,11 +58,54 @@
+    * Stores will return only non expired keys
+    * @implSpec
+    * The default implementation just invokes {@link #publishEntries(Predicate, boolean, boolean)} passing along the
+    * provided filter and <b>false</b> for fetchValue and <b>true</b> for fetchMetadata and retrieves
```
Sure

wburns commented on core/src/main/java/org/infinispan/scattered/impl/ScatteredVersionManagerImpl.java:
```diff
@@ -128,19 +129,16 @@ public void initTopologyId() {
-                  }
-               }
+      Publisher<MarshalledEntry<Object, Object>> publisher = persistenceManager.publishEntries(false, true);
+      Flowable.fromPublisher(publisher).map(MarshalledEntry::getMetadata).blockingForEach(metadata -> {
```
You mean just doing it inside the `blockingForEach` lambda? Not it should be the same.

wburns commented on core/src/main/java/org/infinispan/util/Closeables.java:
```diff
@@ -0,0 +1,48 @@
+      return new CloseableIterator<E>() {
+         @Override
+         public void close() {
+            ((Disposable) iterator).dispose();
```
Only the iterator. The publisher and iterable don't really apply as they don't do anything, until the iterator is invoked. It has all the resources.

wburns commented on core/src/test/java/org/infinispan/persistence/ParallelIterationTest.java:
```diff
@@ -111,18 +116,20 @@ public void testCancelingTaskMultipleProcessors() {
-      }, executor, true, true);
+      Flowable.fromPublisher(loader.publishEntries(null, true, true))
+            .observeOn(Schedulers.from(executor))
+            .takeUntil((MarshalledEntry<Object, Object> me) -> stopped.get())
```
Hrmm tbh I am not exactly sure. I think it has to invoke `doOnNext` before it can call `takeUntil` again. Let me change it and we can find out :)

wburns replied:
Yeah so it will always only modify on a single thread at a time. I can change it to assert it equals the size exactly then. This should be better, because I could make it so we are doing this across multiple threads with `Flowable.parallel` or `Flowable.flatMap`, but then we are testing rxjava2, which isn't the point of this test.

Although tbh you wouldn't want to do cancellation like this in the first place, you should just invoke `Flowable.take(100)` instead.

The more I thought about it, this method should just be removed 🗡

danberindei commented:
Integrated, thanks Will!
